fuse_core/include/fuse_core/transaction.hpp

```diff
@@ -0,0 +1,331 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2018, Locus Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_CORE__TRANSACTION_HPP_
+#define FUSE_CORE__TRANSACTION_HPP_
+
+#include <iostream>
+#include <ostream>
+#include <set>
+#include <vector>
+
+#include <boost/archive/binary_oarchive.hpp>
```

Comment:
Because of the include reordering, I needed to add this in, otherwise one of the boost serialization libraries complains about a missing method.

(The issue used to be masked by fuse_core/serialization.hpp being included first, but that can't happen without making cpplint unhappy.)

Reply:
Sorry. I'm behind on my email.

I just want to note that the boost serialization system can be particular about when/where includes happen. Specifically the input/output archives must be included before any BOOST_CLASS_EXPORT macros are used.

https://www.boost.org/doc/libs/1_71_0/libs/serialization/doc/special.html#export
> BOOST_CLASS_EXPORT in the same source module that includes any of the archive class headers will instantiate code required to serialize polymorphic pointers of the indicated type to the all those archive classes. If no archive class headers are included, then no code will be instantiated.
>
> Note that the implemenation of this functionality requires that the BOOST_CLASS_EXPORT macro appear after the inclusion of any archive class headers for which code is to be instantiated. So, code that uses BOOST_CLASS_EXPORT will look like the following:
>
>```
> #include <boost/archive/text_oarchive.hpp>
> #include <boost/archive/text_oarchive.hpp>
> ... // other archives
> 
> #include "a.hpp" // header declaration for class a
> BOOST_CLASS_EXPORT(a)
> ... // other class headers and exports
> ```